### PR TITLE
Autodetect the number of GPUs when starting Ray.

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Autodetect num gpus when running in local cluster mode. @robertnishihara do you think this should live in tune only or should we do this for all local clusters?

## Related issue number

https://github.com/ray-project/ray/issues/1168

### CODE REVIEW

**Comment by robertnishihara** [comment]
We should probably do it in general. What do you think?

**Comment by ericl** [comment]
Yeah let me make that change.

**Comment by atumanov** [comment]
if we're OK with a dependency on tensorflow (which is already part of installation instructions for Ray), we could use `device_lib.list_local_devices()` from `tensorflow.python.client`. 
This PR would only work on Linux (relies on procfs) and only with NVIDIA GPUs. Are we OK with that?

**Comment by robertnishihara** [comment]
TensorFlow is too heavy weight to depend on. It's not part of the installation instructions at http://ray.readthedocs.io/en/latest/install-on-ubuntu.html.

**Comment by atumanov** [comment]
@robertnishihara , actually those installation instructions are out of date. E.g., missing cloudpickle==0.4.1 . I was referring to tensorflow being the requirement for this:
http://ray.readthedocs.io/en/latest/using-ray-with-tensorflow.html
In any case, it would be good to have a more general solution, if we want to make GPU auto-discovery part of Ray initialization.

**Comment by robertnishihara** [comment]
When you say "more general", you mean "not just NVIDIA GPUs"? Are there other ways in which it would be good to be more general?

**Comment by atumanov** [comment]
Again, the ways in which more generality would be desired:
1. this has dependency on procfs
2. this only supports NVIDIA GPUs [1]. If this is the scope/intention for GPU support to be limited to NVIDIA only, then that's fine.
3. this hard-codes a path to inspect for GPU devices. It would be better to rely on existing discovery libraries or system utilities. I agree that tensorflow dependency is heavy-weight, but perhaps there are other alternatives.

To summarize, wider device support (if in scope), not hardcoding filesystem paths in the code, support for other OSs (again, if in scope).

References:
[1] https://stackoverflow.com/questions/37892784/using-keras-tensorflow-with-amd-gpu

**Comment by atumanov** [comment]
example target platforms and devices (from [1] above):
`it is developed on Ubuntu 16.04 (using Intel HD5500, and NVIDIA GPUs) and Mac Sierra (using Intel HD 530, and Radeon Pro 450)`

**Comment by ericl** [comment]
We can't use tensorflow devicelib since that actually actually initializes
the GPUs when you use it, which takes a long time.

**Comment by ericl** [comment]
Ok, I've changed this to apply to all local ray.init() calls, and in the docs mention that it only works for Nvidia GPUs.

Tested on a p2.xlarge and it seems to work.

**Comment by robertnishihara** [comment]
@ericl I'm going to try moving this code into services.py so that it gets called in the `ray start` code path as well.

**Review by robertnishihara — APPROVED** [review]

**Comment by robertnishihara** [comment]
There was a Travis valgrind failure which looked unrelated. I restarted it.
